webui/src/pages/user/upload-extension-dialog.tsx

```diff
@@ -0,0 +1,273 @@
+        }
+
+        this.setState({ fileToUpload: acceptedFiles[0] });
+    };
+
+    protected handleFileDialogOpen = () => {
+        this.setState({ oldFileToUpload: undefined });
+    };
+
+    protected handleUpload = async () => {
+        if (!this.context.user || !this.state.fileToUpload) {
+            return;
+        }
+
+        this.setState({ uploading: true });
+        let published = false;
+        let retryPublish = false;
+        try {
+            published = await this.tryPublishExtension(this.state.fileToUpload);
+        } catch (err) {
+            retryPublish = await this.tryResolveNamespaceError(err);
+            if (!retryPublish) {
+                this.context.handleError(err);
+            }
+        }
+        if (retryPublish) {
+            try {
+                published = await this.tryPublishExtension(this.state.fileToUpload);
+            } catch (err) {
+                this.context.handleError(err);
+            }
+        }
+        if (published) {
+            this.props.extensionUploaded();
+            this.setState({ open: false });
+        }
+
+        this.setState({ uploading: false });
+    };
+
+    handleEnter = (e: KeyboardEvent) => {
+        if (e.code ===  'Enter') {
+            this.handleUpload();
+        }
+    };
+
+    protected tryPublishExtension = async (fileToUpload: File): Promise<boolean> => {
+        let published = false;
+        const publishResponse = await this.context.service.publishExtension(fileToUpload);
+        if (isError(publishResponse)) {
+            throw publishResponse;
+        }
+
+        published = true;
+        return published;
+    };
+
+    protected tryResolveNamespaceError = async (publishResponse: Readonly<ErrorResult>) => {
+        let resolved = false;
+        try {
+            const namespaceError = 'Unknown publisher: ';
+            if (publishResponse.error.startsWith(namespaceError)) {
+                const namespace = publishResponse.error.substring(namespaceError.length, publishResponse.error.indexOf('\n', namespaceError.length));
+                const namespaceResponse = await this.context.service.createNamespace(namespace);
+                if (isError(namespaceResponse)) {
+                    throw namespaceError;
+                }
+
+                resolved = true;
+            }
+        } catch (err) {
+            this.context.handleError(err);
+        }
+
+        return resolved;
+    };
+
+    componentDidMount() {
+        document.addEventListener('keydown', this.handleEnter);
+    }
+
+    componentWillUnmount() {
+        document.removeEventListener('keydown', this.handleEnter);
+    }
+
+    render() {
+        const classes = this.props.classes;
+        const successClass = this.context.pageSettings.themeType === 'dark' ? classes.successDark : classes.successLight;
+
+        return <React.Fragment>
+            <Button variant='outlined' onClick={this.handleOpenDialog}>Upload extension</Button>
```

Comment:
Maybe we should rethink the wording on these dialogs and buttons. I think if we tell the users that with this button you "upload" an extension, they would think it would not publish it, just upload it as a "draft" or similar. 

I, personally would reword this as "Publish Extension", as well as change the "Upload" button to "Publish" in the modal upload dialog.

Reply:
And maybe also add a little note in the modal, saying something like "After you click 'Publish', this extension will be available on the Marketplace"

Reply:
I've renamed everything from upload to publish.
I added a title to the publish button in the dialog.